File: leetcode/134.cpp
// 机智求解。两个指针逼近。好像和 greedy 没什么关系。

class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int n = gas.size();
        
        int start = 0, end = 0;
        int sum = 0;
        while (1) {
            //cout << "start loop" << endl;
            
            while ( (sum += gas[end] - cost[end]) >= 0) {
                end = (end + 1) % n;
                if (end == start)
                    return start;
            }
            //cout << "stop at start = " << start << " end = " << end << " sum = " << sum << endl;

            while (sum < 0) {
                if (--start < 0)
                    start = n - 1;
                if (start == end)
                    return sum >= 0 ? start : -1;
                sum += gas[start] - cost[start];
                //cout << "start = " << start << " end = " << end << " sum = " << sum << endl;
            }
            end++;
        }
    }
};

File: leetcode/134.rs
// 和买股票题是一样的。
impl Solution {
    pub fn can_complete_circuit(gas: Vec<i32>, cost: Vec<i32>) -> i32 {
        let n = gas.len();
        let mut i = 0;
        let mut i_step = 0;
        while i_step < n {
            let mut j = i;
            let mut left = gas[i] - cost[i];
            let mut j_step = 0;
            while left >= 0 {
                j = (j + 1) % n;
                left += gas[j] - cost[j];
                j_step += 1;
                if j_step == n {
                    return i as i32;
                }
            }
            i = (j + 1) % n;
            i_step += j_step + 1;
        }
        return -1;
    }
}
